fault-proof/tests/common/contracts.rs

```diff
@@ -0,0 +1,263 @@
+
+    // Deploy factory proxy with initialization
+    debug!("Deploying DisputeGameFactory proxy...");
+    let init_data = DisputeGameFactory::initializeCall { _owner: DEPLOYER_ADDRESS }.abi_encode();
+
+    let factory_proxy =
+        ERC1967Proxy::deploy(provider.clone(), factory_impl_addr, init_data.into()).await?;
+    let factory = *factory_proxy.address();
+    info!("✓ DisputeGameFactory proxy deployed at: {}", factory);
+
+    // 2. Deploy MockOptimismPortal2 (needed for AnchorStateRegistry)
+    debug!("Deploying MockOptimismPortal2...");
+    let portal_instance =
+        MockOptimismPortal2::deploy(provider.clone(), TEST_GAME_TYPE, U256::from(AIRGAP_PERIOD))
+            .await?;
+    let portal = *portal_instance.address();
+    info!("✓ MockOptimismPortal2 deployed at: {}", portal);
+
+    // 3. Deploy AnchorStateRegistry as proxy (before AccessManager)
+    debug!("Deploying SuperchainConfig...");
+    let superchain_config = SuperchainConfig::deploy(provider.clone()).await?;
+    let superchain_config_addr = *superchain_config.address();
+    info!("✓ SuperchainConfig deployed at: {}", superchain_config_addr);
+
+    // Deploy AnchorStateRegistry implementation
+    debug!("Deploying AnchorStateRegistry implementation...");
+    let anchor_impl = AnchorStateRegistry::deploy(provider.clone()).await?;
+    let anchor_impl_addr = *anchor_impl.address();
+    info!("✓ AnchorStateRegistry implementation deployed at: {}", anchor_impl_addr);
+
+    // Prepare initialization data with expected test values
+    // NOTE: We use the finalized L2 block number - 100 as the starting anchor root and test 3 games
+    // with proposal interval of 10 blocks.
+    let l2_block_number = U256::from(
+        l2_provider.get_l2_block_by_number(BlockNumberOrTag::Finalized).await?.header.number,
+    ) - U256::from(100u64);
+    let output_root = l2_provider.compute_output_root_at_block(l2_block_number).await?;
+    let starting_anchor_root = bindings::anchor_state_registry::AnchorStateRegistry::OutputRoot {
+        root: output_root,
+        l2BlockNumber: l2_block_number,
+    };
+
+    let init_data = bindings::anchor_state_registry::AnchorStateRegistry::initializeCall {
+        _superchainConfig: superchain_config_addr,
+        _disputeGameFactory: factory,
+        _portal: portal,
+        _startingAnchorRoot: starting_anchor_root,
+    }
+    .abi_encode();
+
+    // Deploy AnchorStateRegistry proxy
+    debug!("Deploying AnchorStateRegistry proxy...");
+    let anchor_proxy =
+        ERC1967Proxy::deploy(provider.clone(), anchor_impl_addr, init_data.into()).await?;
+    let anchor_state_registry = *anchor_proxy.address();
+    info!("✓ AnchorStateRegistry proxy deployed at: {}", anchor_state_registry);
+
+    // 4. Deploy AccessManager
+    debug!("Deploying AccessManager...");
+    // Note: AccessManager deployment may fail on Anvil forks, so we handle it gracefully
```

Comment:
why does this fail and anchor state registry not fail?

Reply:
For the AccessManager, the problem is a combination of two things:
1. Ownable Contract: AccessManager inherits OpenZeppelin's Ownable, which sets the contract owner to msg.sender inside its constructor.
2. Anvil Fork Bug: Anvil's gas estimator struggles to correctly calculate the gas needed for this specific type of constructor logic when running on a forked network.

AnchorStateRegistry succeeds because it follows the proxy pattern. Its core logic isn't in a constructor but in an initialize function that is called after deployment. This basic proxy deployment is handled correctly with Anvil's gas estimator.